o/servicestate/quota_control.go: introduce basic group manipulation methods

anonymouse64 opened the pull request:
cc @stolowski these methods (or something very close to them hopefully) are what the daemon backend will need to use

I still need to write tests for these methods, but wanted to get alignment on the set of functions here... Also was wondering if we should be creating tasks / changes for these operations to match i.e. `snap set` and `snap start`. Arguably we could, but also maybe we want to avoid that complexity.

Based on top of https://github.com/snapcore/snapd/pull/10153

stolowski commented on overlord/servicestate/quota_control.go:
```diff
@@ -0,0 +1,232 @@
+	return nil
+}
+
+func (mgr *ServiceManager) AddSnapToQuota(grpName string, snapName string) error {
```
Would it make sense to take multiple snap names to 1) ensure all preconditions first for all and 2) call ensureSnapServicesForGroup only once for the given group?

anonymouse64 replied:
yeah this is fixed now

stolowski commented:
@anonymouse64 What's the idea for updating quota groups? The problem I see with proposed approach is there is no way to update the given group at once (i.e. append a bunch of snaps and a new memory limit as a single op). I'd expect to be able to update the state first, then have evertything synced in systemd?

stolowski commented on overlord/servicestate/quota_control.go:
```diff
@@ -0,0 +1,232 @@
+	return nil
+}
+
+func (mgr *ServiceManager) ChangeGroupMemoryLimit(grpName string, memoryLimit quantity.Size) error {
```
This is related to the previous question and the general one I asked under the PR; I think we should have a way to update all aspects of the group at once.

anonymouse64 replied:
this has been refactored to use a more generic UpdateQuota method which hopefully is as generic as we need it to be

anonymouse64 commented:
I had to expand the scope of this PR a bit to get things working, mainly I added these things:
- [x] added wrappers.RemoveQuotaGroup
- [x] fixed up snap/quota package to work with orphaning or adopting sub-groups
- [x] addressed feedback from Michael on pre-req that landed
- [x] renamed UpdateQuotas -> PatchQuotasState so it doesn't get confused with UpdateQuota that I added as a backend method for daemon to use

The only things still missing from this PR at my EOD on Wednesday are:

- [ ] unit test about groups already existing when calling CreateQuota
- [ ] unit test about group not already existing when calling RemoveQuota

also there are not yet any unit tests about UpdateQuota, but specifically there should be:
- [ ] group being updated does not exist
- [ ] invalid option set provided (i.e. orphaning a non-sub-group, etc.)
- [ ] orphaning a valid sub-group
- [ ] moving a valid sub-group from one parent to another
- [ ] adopting a non-sub-group to a parent
- [ ] simple mutation like changing memory limit
- [ ] full replacement of the set of snaps in a quota
- [ ] appending to the set of snaps in a quota

mvo5 reviewed (commented):
This is great but at this point it seems a bit more than what we actually need for the MVP we need for Friday. We will use this (or a variant) of course but for today I hope we can get away with a slightly simplified version that I pushed to #10211 - that is limited to add/remove but you are much more into this topic so hopefully I did not do something silly there.

pedronis commented:
marking blocked mostly to clarify that the entirety of this is not our focus until we get a MVP based on simpler things first (see #10211)
